contracts/schemes/ArcUpgradeScheme.sol

```diff
@@ -0,0 +1,200 @@
+ */
+contract ArcUpgradeScheme is VotingMachineCallbacks, ProposalExecuteInterface, Initializable {
+    event NewUpgradeProposal(
+        address indexed _avatar,
+        bytes32 indexed _proposalId,
+        uint64[3] _packageVersion,
+        bytes32[] _contractsNames,
+        address[] _contractsToUpgrade,
+        string  _descriptionHash
+    );
+
+    event ProposalExecuted(
+        address indexed _avatar,
+        bytes32 indexed _proposalId
+    );
+
+    event ProposalExecutedByVotingMachine(
+        address indexed _avatar,
+        bytes32 indexed _proposalId,
+        int256 _param
+    );
+
+    event ProposalDeleted(address indexed _avatar, bytes32 indexed _proposalId);
+
+    // Details of a voting proposal:
+    struct UpgradeProposal {
+        uint64[3] packageVersion;
+        bytes32[] contractsNames;
+        address[] contractsToUpgrade;
+        bool exist;
+        bool passed;
+    }
+
+    mapping(bytes32=>UpgradeProposal) public organizationProposals;
+
+    IntVoteInterface public votingMachine;
+    bytes32 public voteParams;
+    Avatar public avatar;
+    Package public package;
+
+    /**
+     * @dev initialize
+     * @param _avatar the avatar to mint reputation from
+     * @param _votingMachine the voting machines address to
+     * @param _voteParams voting machine parameters.
+     */
+    function initialize(
+        Avatar _avatar,
+        IntVoteInterface _votingMachine,
+        bytes32 _voteParams,
+        Package _package
+    )
+    external
+    initializer
+    {
+        require(_avatar != Avatar(0), "avatar cannot be zero");
+        avatar = _avatar;
+        votingMachine = _votingMachine;
+        voteParams = _voteParams;
+        package = _package;
+    }
+
+    /**
+    * @dev execution of proposals, can only be called by the voting machine in which the vote is held.
+    * @param _proposalId the ID of the voting in the voting machine
+    * @param _decision a parameter of the voting result, 1 yes and 2 is no.
+    * @return bool success
+    */
+    function executeProposal(bytes32 _proposalId, int256 _decision)
+    external
+    onlyVotingMachine(_proposalId)
+    returns(bool) {
+        UpgradeProposal storage proposal = organizationProposals[_proposalId];
+        require(proposal.exist, "must be a live proposal");
+        require(proposal.passed == false, "cannot execute twice");
+
+        if (_decision == 1) {
+            proposal.passed = true;
+            execute(_proposalId);
+        } else {
+            delete organizationProposals[_proposalId];
+            emit ProposalDeleted(address(avatar), _proposalId);
+        }
+
+        emit ProposalExecutedByVotingMachine(address(avatar), _proposalId, _decision);
+        return true;
+    }
+
+    /**
+    * @dev execution of proposals after it has been decided by the voting machine
+    * @param _proposalId the ID of the voting in the voting machine
+    */
+    function execute(bytes32 _proposalId) public {
```

Comment:
why this is public function ?
is there a chance that the genericCall will revert on the execution time and will not refer after it ?

Reply:
Only chance I think is if the Avatar is not the admin of the contract proposed. Not sure it’s necessary to have this public but I just sticked to the same way you made GenericScheme

Reply:
it should not be public

Reply:
You mean here or also in the GenericScheme?

Reply:
here. 
in GenericScheme it is different case where there is a chance that the target contract call will revert though the arc tx will not due to the nature of contract.call ...

The case in this scheme is different as there is no way the target contract can revert.

Reply:
ok.